runtime/integration-tests/src/cases/lp/utils.rs

```diff
@@ -0,0 +1,310 @@
+		pallet_liquidity_pools_gateway::Event::<T>::OutboundMessageExecutionFailure { .. }
+	));
+}
+
+pub fn verify_outbound_success<T: Runtime>(
+	message: <T as pallet_liquidity_pools_gateway::Config>::Message,
+) {
+	assert!(matches!(
+		last_event::<T, pallet_liquidity_pools_gateway::Event::<T>>(),
+		pallet_liquidity_pools_gateway::Event::<T>::OutboundMessageExecutionSuccess {
+			message: processed_message,
+			..
+		} if processed_message == message
+	));
+}
+
+pub fn process_outbound<T: Runtime>(
+	mut verifier: impl FnMut(<T as pallet_liquidity_pools_gateway::Config>::Message),
+) {
+	let msgs = pallet_liquidity_pools_gateway::OutboundMessageQueue::<T>::iter()
+		.map(|(nonce, (_, _, msg))| (nonce, msg))
+		.collect::<Vec<_>>();
+
+	// The function should panic if there is nothing to be processed.
+	assert!(msgs.len() > 0);
+
+	msgs.into_iter().for_each(|(nonce, msg)| {
+		pallet_liquidity_pools_gateway::Pallet::<T>::process_outbound_message(
+			OriginFor::<T>::signed(Keyring::Alice.into()),
+			nonce,
+		)
+		.unwrap();
+
+		verifier(msg);
+	});
+}
+
+pub fn to_fixed_array<const S: usize>(src: &[u8]) -> [u8; S] {
+	let mut dest = [0; S];
+	let len = min(src.len(), S);
+	dest[..len].copy_from_slice(&src[..len]);
+
+	dest
+}
+
+pub fn as_h160_32bytes(who: Keyring) -> [u8; 32] {
+	let mut address = [0u8; 32];
+	address[..20].copy_from_slice(H160::from(who).as_bytes());
+	address
+}
+
+trait Input {
+	fn input(&self) -> &[u8];
+}
+
+impl Input for Vec<u8> {
+	fn input(&self) -> &[u8] {
+		self.as_slice()
+	}
+}
+
+impl<E: Debug> Input for Result<Vec<u8>, E> {
+	fn input(&self) -> &[u8] {
+		match self {
+			Ok(arr) => arr.as_slice(),
+			Err(e) => panic!("Input received error: {:?}", e),
+		}
+	}
+}
+
+impl<E: Debug> Input for Result<ExecutionInfo<Vec<u8>>, E> {
+	fn input(&self) -> &[u8] {
+		match self {
+			Ok(arr) => arr.value.as_slice(),
+			Err(e) => panic!("Input received error: {:?}", e),
+		}
+	}
+}
+
+pub trait Decoder<T> {
+	fn decode(&self) -> T;
+}
+
+impl<T: Input> Decoder<(bool, u64)> for T {
+	fn decode(&self) -> (bool, u64) {
+		assert!(self.input().len() > 32);
+
+		let left = &self.input()[..32];
+		let right = &self.input()[32..];
+
+		let unsigned64 = match right.len() {
+			1 => u64::from(u8::from_be_bytes(to_fixed_array(&right))),
+			2 => u64::from(u16::from_be_bytes(to_fixed_array(&right))),
+			4 => u64::from(u32::from_be_bytes(to_fixed_array(&right))),
+			8 => u64::from_be_bytes(to_fixed_array(&right)),
+			// EVM stores in 32 byte slots with left-padding
+			16 => u64::from_be_bytes(to_fixed_array::<8>(&right[28..])),
+			32 => u64::from_be_bytes(to_fixed_array::<8>(&right[24..])),
+			_ => {
+				panic!("Invalid slice length for u64 derivation");
+			}
+		};
+
+		(left[31] == 1u8, unsigned64)
+	}
+}
+
+impl<T: Input> Decoder<H160> for T {
+	fn decode(&self) -> H160 {
+		assert_eq!(self.input().len(), 32usize);
+
+		H160::from(to_fixed_array(&self.input()[12..]))
```

Comment:
Our `convert_evm_address()` starts with the H160 and the last 12 are the not used. But here it is the opposite

Reply:
Good catch. Running tests now over lunch break to see whether this breaks any ITs.

Reply:
To be honest, I'm not sure why in some Decode Methods, it is expected to be 32 bytes, but some bytes are skipped, such as when decoding the `Balance`.

Reply:
Yeah it is quite magic indeed and I also don't fully understand it. Regarding this code path, it seems to be correct. At least it has been like this since its genesis when Frederik initialized LP integration tests ([blame](https://github.com/centrifuge/centrifuge-chain/blame/e4e048a574e393061f0169e88d2d66b7e1970760/runtime/integration-tests/src/generic/cases/lp/mod.rs#L244)).

It is used for decoding addresses returned by the byte vector `evm.view(...).unwrap().value`: https://github.com/centrifuge/centrifuge-chain/blob/020cf59f18209768afbb8a844867b976f66ab65d/runtime/integration-tests/src/cases/lp/pool_management.rs#L90

Under the hood, it is the result of https://github.com/centrifuge/centrifuge-chain/blob/e95db3922e81e371571cb899e3c80bfcac255f4c/runtime/integration-tests/src/envs/evm_env.rs#L181

So it appears to me that `pallet_evm` prefixes 20-byte addresses with zeroes. After changing the above line 
```diff
- 		H160::from(to_fixed_array(&self.input()[12..]))
+ 		H160::from(to_fixed_array(&self.input()[..20]))
```

many tests fail with weird reasons. So let's keep as is.

Reply:
Ok, it seems like all call answers from Ethereum return 32 bytes. And in those bytes there are coded different types. So always need to be 32 bytes and the standard of this will dictate how a H160 is represented in those 32 bytes. Thanks for the clarification!